compiler/rustc_typeck/src/collect/type_of.rs

```diff
@@ -483,15 +512,59 @@ pub(super) fn type_of(tcx: TyCtxt<'_>, def_id: DefId) -> Ty<'_> {
                     .discr_type()
                     .to_ty(tcx),
 
+                Node::TraitRef(trait_ref @ &TraitRef {
+                  path, ..
+                }) if let Some((binding, seg)) =
+                  path
+                      .segments
+                      .iter()
+                      .filter_map(|seg| seg.args.map(|args| (args.bindings, seg)))
+                      .find_map(|(bindings, seg)| {
+                          bindings
+                              .iter()
+                              .filter_map(|binding| {
+                                  if let hir::TypeBindingKind::Equality { term: Term::Const(c) } =
+                                      binding.kind
+                                  {
+                                          Some((binding, c))
+                                  } else {
+                                      None
+                                  }
+                              })
+                              .find_map(|(binding, ct)| if ct.hir_id == hir_id {
+                                Some((binding, seg))
+                              } else {
+                                None
+                              })
+                      }) =>
+                {
+                  // FIXME(associated_const_equality) when does this unwrap fail? I have no idea what case it would.
+                  let trait_def_id = trait_ref.trait_def_id().unwrap();
+                  let assoc_items = tcx.associated_items(trait_def_id);
+                  let assoc_item = assoc_items.find_by_name_and_kind(
+                    tcx, binding.ident, ty::AssocKind::Const, def_id.to_def_id(),
+                  );
+                  if let Some(assoc_item) = assoc_item {
+                    tcx.type_of(assoc_item.def_id)
+                  } else {
+                      // FIXME(associated_const_equality): add a useful error message here.
```

Comment:
have you been able to trigger this message?

Reply:
Nope, I'll add a test for it

Reply:
I don't think this would get hit, unsure when it could

Reply:
That seems to be the case for multiple errors reported in this file, but your code matches the ongoing theme, so let's stick with it. Just remove the FIXME then